### ISSUE TITLE
reduce() of empty sequence with no initial value

### ISSUE THREAD

**Issue by cy69855522** [zalandoresearch/flair#589]
I modify the text classification demo to use text cnn but get this error.
`reduce() of empty sequence with no initial value` in training_utils.py WeightExtractor.
I use stacked embedding to replace document embedding and change the model in TextClassifier.
And get this error T W T .
Could you please help me to look for the reason？

"""=====================================================================================================================
5 create the text classifier
"""
from typing import List, Union
from flair.nn import LockedDropout, WordDropout
from flair.data import Dictionary, Sentence

class ResnetBlock(nn.Module):
    def __init__(self, channel_size):
        super(ResnetBlock, self).__init__()
        self.channel_size = channel_size
        self.maxpool = nn.Sequential(
            nn.ConstantPad1d(padding=(0, 1), value=0),
            nn.MaxPool1d(kernel_size=3, stride=2)
        )
        self.conv = nn.Sequential(
            nn.BatchNorm1d(num_features=self.channel_size),
            nn.ReLU(),
            nn.Conv1d(self.channel_size, self.channel_size,
                      kernel_size=3, padding=1),
            nn.BatchNorm1d(num_features=self.channel_size),
            nn.ReLU(),
            nn.Conv1d(self.channel_size, self.channel_size,
                      kernel_size=3, padding=1),
        )

    def forward(self, x):
        x_shortcut = self.maxpool(x)
        x = self.conv(x_shortcut)
        x = x + x_shortcut
        return x

class MyClassifier(TextClassifier):
    def __init__(self, embeddings: StackedEmbeddings,
                 label_dictionary: Dictionary,
                 multi_label: bool,
                 max_sentence_len: int = 40,
                 dropout: float = 0.5,
                 opt = None
                ):
        
        
        super(TextClassifier, self).__init__()
        
        self.embeddings = embeddings
        self.label_dictionary: Dictionary = label_dictionary
        self.multi_label = multi_label
        self.max_sentence_len = max_sentence_len
        self.embeddings_dimension = self.embeddings.embedding_length
        self.dropout = dropout
        
        # main trunk
        question_convs1 = [nn.Sequential(
                nn.Conv1d(in_channels=self.embeddings_dimension,
                          out_channels=opt.TITLE_DIM,
                          kernel_size=kernel_size),
                nn.BatchNorm1d(opt.TITLE_DIM),
                nn.ReLU(inplace=True),

                nn.MaxPool1d(kernel_size=(opt.SENT_LEN - kernel_size + 1))
            )for kernel_size in opt.SIN_KER_SIZE]

        question_convs2 = [nn.Sequential(
                nn.Conv1d(in_channels=self.embeddings_dimension,
                          out_channels=opt.TITLE_DIM,
                          kernel_size=kernel_size[0]),
                nn.BatchNorm1d(opt.TITLE_DIM),
                nn.ReLU(inplace=True),

                nn.Conv1d(in_channels=opt.TITLE_DIM,
                          out_channels=opt.TITLE_DIM,
                          kernel_size=kernel_size[1]),
                nn.BatchNorm1d(opt.TITLE_DIM),
                nn.ReLU(inplace=True),
                nn.MaxPool1d(kernel_size=(opt.SENT_LEN - kernel_size[0] - kernel_size[1] + 2))
            )for kernel_size in opt.DOU_KER_SIZE]

        question_convs = question_convs1
        question_convs.extend(question_convs2)

        self.num_seq = len(opt.DOU_KER_SIZE) + len(opt.SIN_KER_SIZE)
        self.change_dim_conv  = nn.Conv1d(opt.TITLE_DIM, opt.NUM_ID_FEATURE_MAP, kernel_size=1, stride=1)

        self.question_convs = nn.ModuleList(question_convs)
        self.conv = nn.Sequential(
            nn.BatchNorm1d(num_features=opt.TITLE_DIM),
            nn.ReLU(),
            nn.Conv1d(opt.TITLE_DIM, opt.NUM_ID_FEATURE_MAP,
                      kernel_size=3, padding=1),
            nn.BatchNorm1d(num_features=opt.NUM_ID_FEATURE_MAP),
            nn.ReLU(),
            nn.Conv1d(opt.NUM_ID_FEATURE_MAP, opt.NUM_ID_FEATURE_MAP,
                      kernel_size=3, padding=1)
        )
        self.blocks = []
        while self.num_seq > 2:
            self.blocks.append(ResnetBlock(opt.NUM_ID_FEATURE_MAP).cuda())
            self.num_seq //= 2
        self.fc = nn.Sequential(
            nn.Linear(opt.NUM_ID_FEATURE_MAP*self.num_seq, len(self.label_dictionary)),
            nn.BatchNorm1d(len(self.label_dictionary)),
            nn.ReLU(inplace=True),
            nn.Dropout(self.dropout),
            nn.Linear(len(self.label_dictionary), len(self.label_dictionary))
        )
        
        
        
        # init weights
        self._init_weights()

        if multi_label:
            self.loss_function = nn.BCELoss()
        else:
            self.loss_function = nn.CrossEntropyLoss()

        # auto-spawn on GPU if available
        self.to(flair.device)
        
    def _init_weights(self):
        pass

    def forward(self, sentences: Union[List[Sentence], Sentence]) -> List[List[float]]:
        if type(sentences) is Sentence:
            sentences = [sentences]
        
        self.embeddings.embed(sentences)
        all_sentence_tensors = []

        # go through each sentence in batch
        for i, sentence in enumerate(sentences):
            
            word_embeddings = []
            for j, token in enumerate(sentence.tokens):
                word_embeddings.append(token.get_embedding().unsqueeze(0))
                if j == self.max_sentence_len:
                    break

            # PADDING: pad shorter sentences out
            for j in range(self.max_sentence_len - len(sentence.tokens)):
                word_embeddings.append(
                    torch.zeros(self.embeddings_dimension,
                                dtype=torch.float).unsqueeze(0)
                )
                
            word_embeddings_tensor = torch.cat(word_embeddings, 0).to(flair.device)

            # ADD TO SENTENCE LIST: add the representation
            all_sentence_tensors.append(word_embeddings_tensor.unsqueeze(0))

        # --------------------------------------------------------------------
        # GET REPRESENTATION FOR ENTIRE BATCH
        # --------------------------------------------------------------------
        sentence_tensor = torch.cat(all_sentence_tensors, 0) # batch_size, sentence_len, w2v_size

        # --------------------------------------------------------------------
        # FF PART
        # --------------------------------------------------------------------
        

        x = [question_conv(sentence_tensor.permute(0, 2, 1)) # batch_size, w2v_size, sentence_len
             for question_conv in self.question_convs]
        x = torch.cat(x, dim=2)
        xp = x
        xp = self.change_dim_conv(xp)
        x = self.conv(x)
        x = x+xp
        for block in self.blocks:
            x1 = block(x)
            x = list(block.children())[0](x)
            x = x + x1
        x = x.view(x.shape[0], -1)
        x = self.fc(x)
        return x
    
class obj(): 
    def __init__(self):
        self.USE_CUDA = True
        self.TITLE_DIM = 200 # 中间层维度
        self.SENT_LEN = 40 # 句子截断长度
        self.KERNEL_SIZE = [1,2,3,4,5] # 卷积核大小
        self.NUM_ID_FEATURE_MAP = 250
        self.SIN_KER_SIZE = [1, 3] # 单层卷积卷积核大小 
        self.DOU_KER_SIZE = [(1, 3), (3, 5)] # 双层卷积卷积核大小。元组内第一项为第一层 # conv的核尺寸，第二项为第二层conv的核尺寸
        
opt = obj()
classifier = MyClassifier(stacked_embeddings, label_dictionary=label_dict, multi_label=False, opt=opt)
trainer = ModelTrainer(classifier, corpus, optimizer=AdaBound)
